### DISCUSSION TITLE
Multiple Server

### DISCUSSION THREAD

**Discussion by LinusLing0910** [pymodbus-dev/pymodbus#1789] [Help] [answered]
Is there any example to start multiple tcp server and stop the server one by one  for pymodbus 3.5.2?

I used multiple threads to start tcp server with different port, and call serverstop() to close the server, but it only closed the last active one, but others cannot be stop.

**Comment by janiversen** [comment]
Using multiple threads is asking for trouble, mixing threads and asyncio normally causes problem.

Use the async classes (start/stop methods) they can run multiple servers out of the box.

There are no specific example, but any server started with StartAsync can run in parallel automatically.

**Reply by LinusLing0910** [reply]
I also tried asyncio.run(StartAsyncTcpServer(***)), but it blocked main thread, I just want to start multiple servers in the backgroud (which don't blocked the main thread) and use client to connect different server at one time, and then stop the server one by one

**Reply by janiversen** [reply, marked as answer]
It only blocks the thread if you do await, but look in the code then you can see how we do it.
